magma/clock.py

```diff
@@ -90,17 +90,31 @@ def ClockInterface(has_enable=False, has_reset=False, has_ce=False,
 
 
 def _wire_clock_port(port, clocktype, defnclk):
-    if isinstance(port, (Array, Tuple)):
+    wired = False
+    if isinstance(port, Tuple):
         for elem in port:
-            _wire_clock_port(elem, clocktype, defnclk)
-        return
-    if isinstance(port, clocktype) and not port.driven():
+            wired |= _wire_clock_port(elem, clocktype, defnclk)
+    elif isinstance(port, Array):
+        wired = _wire_clock_port(port[0], clocktype, defnclk)
+        # Only traverse all children circuit if first child has a clock
+        if wired:
+            for elem in port[1:]:
+                _wire_clock_port(elem, clocktype, defnclk)
+    elif isinstance(port, clocktype) and not port.driven():
         wire(defnclk, port)
+        wired = True
+    return wired
 
 
 def _get_clocks(port, clocktype):
-    if isinstance(port, (Array, Tuple)):
-        return sum([_get_clocks(elem, clocktype) for elem in port], [])
+    if isinstance(port, Tuple):
+        return sum((_get_clocks(elem, clocktype) for elem in port), [])
+    if isinstance(port, Array):
+        result = _get_clocks(port[0], clocktype)
+        # Only traverse all children circuit if first child has a clock
+        if result:
+            result += sum((_get_clocks(elem, clocktype) for elem in port[1:]), [])
+        return result
```

Comment:
oh another option here is to make sure the common case is fast. presumably the common case is connecting to a top-level clock port (not a clock in a nested type). We could first check if any port is itself an instance of `clocktype` before doing the recursive checks.

maybe in `wireclocktype` the first few lines should be:

```python
clks = (port if isinstance(port, clocktype) else None
        for port in defn.interface.ports.values())
defnclk = _first(clks)
if defnclk is None:
  # Call _get_first_clock() as described in above comment.
```